### PR TITLE
Create API projects via apictl

### PR DESCRIPTION
**Pull request by SuKSW**
### Purpose
- Enable creating API projects via apictl for all test cases
- Deploy API via apictl in all test cases
- Add apictl testcases
- Use test tags in testng.xml to reuse mgw instances
- Move a few util methods to where similar methods exist
- Create util methods to be reused
- Move all test cases into the same folder
- Add deleteAPI testcase
- Rename test module to integration

### Issues
Closes https://github.com/wso2/product-microgateway/issues/1709
Closes https://github.com/wso2/product-microgateway/issues/1735

### Automation tests
 - Unit tests added: Yes/No
 - Integration tests added: Yes

### Tested environments
OS name: "linux", arch: "amd64", family: "unix"
OpenJDK Runtime Environment (build 11.0.10+9-Ubuntu-0ubuntu1.20.04)

---
#### Maintainers: Check before merge
- [x] Assigned 'Type' label
- [x] Assigned the project
- [x] Validated respective github issues
- [x] Assigned milestone to the github issue(s)

### CODE REVIEW

File: test/test-integration/pom.xml [hunk 1]
```diff
@@ -108,6 +108,24 @@
...
+                <artifactId>download-maven-plugin</artifactId>
+                <executions>
+                    <execution>
+                        <id>download-apictl</id>
+                        <phase>generate-resources</phase>
+                        <goals>
+                            <goal>wget</goal>
+                        </goals>
+                        <configuration>
+                            <url>https://wso2.org/jenkins/job/platform-builds/job/product-apim-tooling/lastSuccessfulBuild/artifact/archive/${apictl.package}</url>
```
Comment (SuKSW) [inline test/test-integration/pom.xml:122]: Change to get from github release

File: test/test-integration/pom.xml [hunk 2]
```diff
@@ -108,6 +108,24 @@
...
+                            <unpack>true</unpack>
+                            <outputDirectory>${basedir}/../test-integration/target/</outputDirectory>
```
Comment (SuKSW) [inline test/test-integration/pom.xml:124]: Check property to directly get target dir

File: test/test-integration/pom.xml [hunk 3]
```diff
@@ -162,7 +180,7 @@
         <dependency>
             <groupId>org.apache.commons</groupId>
             <artifactId>commons-lang3</artifactId>
-            <scope>test</scope>
+<!--            <scope>test</scope>-->
```
Comment (SuKSW) [inline test/test-integration/pom.xml:183]: uncomment this

File: test/test-integration/pom.xml [hunk 4]
```diff
@@ -204,4 +222,94 @@
...
+                    <family>windows</family>
+                    <arch>x86_64</arch>
+                </os>
+            </activation>
+            <properties>
+                <apictl.package>apictl-${apictl.version}-windows-x64.zip</apictl.package>
+            </properties>
+        </profile>
+        <profile>
+            <id>windows-i586</id>
```
Comment (SuKSW) [inline test/test-integration/pom.xml:299]: Check in the apictl code

File: test/test-integration/src/test/java/org/wso2am/micro/gw/tests/common/BaseTestCase.java [hunk 5]
```diff
@@ -102,82 +86,5 @@ public void startMGW(String confPath, boolean tlsEnabled, boolean customJwtTrans
      */
     public void stopMGW() {
         microGWServer.stopMGW();
-
-    }
-
-    public static String getImportAPIServiceURLHttps(String servicePath) throws MalformedURLException {
-        return new URL(new URL("https://localhost:" + TestConstant.ADAPTER_IMPORT_API_PORT), servicePath)
```
Comment (SuKSW) [inline test/test-integration/src/test/java/org/wso2am/micro/gw/tests/common/BaseTestCase.java:109]: change URLs file content to HttpClientRequest

File: test/test-integration/src/test/java/org/wso2am/micro/gw/tests/testCaseBefore/MgwWithDefaultConf.java [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+
+import java.io.IOException;
+import java.util.concurrent.TimeUnit;
+
+public class MgwWithDefaultConf extends BaseTestCase {
+
+    @BeforeSuite
+    public void checkVersion() throws IOException {
+        String versionByApictl = ApictlUtils.getVersion();
+        Assert.assertEquals(versionByApictl, "4.0.0-alpha2","Expected apictl version is not downloaded");
```
Comment (SuKSW) [inline test/test-integration/src/test/java/org/wso2am/micro/gw/tests/testCaseBefore/MgwWithDefaultConf.java:36]: Get from pom

File: test/test-integration/src/test/java/org/wso2am/micro/gw/tests/testCases/apiDeploy/APiDeployViaApictlTestCase.java [hunk 7]
```diff
@@ -0,0 +1,36 @@
...
+    public void createApiProject() throws IOException, MicroGWTestException {
+        ApictlUtils.createProject( "deploy_openAPI.yaml",
+                "apictl_petstore", null);
+        ApictlUtils.addEnv("test");
+    }
+
+    @Test
+    public void deployAPI() throws MicroGWTestException {
+        ApictlUtils.login("test");
+        ApictlUtils.deployAPI("apictl_petstore", "test");
```
Comment (SuKSW) [inline test/test-integration/src/test/java/org/wso2am/micro/gw/tests/testCases/apiDeploy/APiDeployViaApictlTestCase.java:23]: Update all tests to deploy via apictl

File: test/test-integration/src/test/java/org/wso2am/micro/gw/tests/testCases/endpoints/ProductionSandboxTestCase.java [hunk 8]
```diff
@@ -16,71 +16,46 @@
...
 
     @Test(description = "Invoke Production and Sandbox endpoint when both endpoints provided")
     public void invokeProdSandEndpoints() throws Exception {
         Map<String, String> prodHeaders = new HashMap<String, String>();
         prodHeaders.put(HttpHeaderNames.AUTHORIZATION.toString(), "Bearer " + jwtTokenProd);
-        HttpResponse prodResponse = HttpsClientRequest.doGet(getServiceURLHttps(
-                "/v2/pet/findByStatus") , prodHeaders);
+        HttpResponse prodResponse = HttpsClientRequest.doGet(URLs.getServiceURLHttps(
+                "/v2/general/pet/findByStatus") , prodHeaders);
+        System.out.println("heyy");
```
Comment (SuKSW) [inline test/test-integration/src/test/java/org/wso2am/micro/gw/tests/testCases/endpoints/ProductionSandboxTestCase.java:55]: Remove these

File: test/test-integration/src/test/java/org/wso2am/micro/gw/tests/testCases/security/CorsTestCase.java [hunk 9]
```diff
@@ -57,38 +51,18 @@
 
     @BeforeClass(description = "initialise the setup")
     void start() throws Exception {
-        super.startMGW(null, true);
-
-        //deploy the api
-        //api yaml file should put to the resources/apis/openApis folder
-        String apiZipfile = ApiProjectGenerator.createApictlProjZip("cors/api.yaml", "cors/swagger.yaml");
+        String apiZipfile = ApictlUtils.createProjectZip( "cors_openAPI.yaml",
+                "cors_petstore", null);
```
Comment (SuKSW) [inline test/test-integration/src/test/java/org/wso2am/micro/gw/tests/testCases/security/CorsTestCase.java:55]: Move to before tests

File: test/test-integration/src/test/java/org/wso2am/micro/gw/tests/util/ApiDeployment.java [hunk 10]
```diff
@@ -49,16 +50,17 @@ public static void deployAPI(String apiZipFilePath) throws Exception {
         // Set header
         Map<String, String> headers = new HashMap<String,String>();
         headers.put(HttpHeaderNames.AUTHORIZATION.toString(), "Basic YWRtaW46YWRtaW4=");
-        HttpsPostMultipart multipart = new HttpsPostMultipart(BaseTestCase.getImportAPIServiceURLHttps(
-                TestConstant.ADAPTER_IMPORT_API_RESOURCE) , headers);
+        HttpsPostMultipart multipart = new HttpsPostMultipart(URLs.getAdapterServiceURLHttps(
```
Comment (SuKSW) [inline test/test-integration/src/test/java/org/wso2am/micro/gw/tests/util/ApiDeployment.java:53]: Remove this class

File: test/test-integration/src/test/java/org/wso2am/micro/gw/tests/util/ApictlUtils.java [hunk 11]
```diff
@@ -0,0 +1,190 @@
...
+    }
+
+    public static String createProjectZip(String openApiFile, String apiProjectName, String backendCert) throws IOException, MicroGWTestException {
+        String apiProjectPath = "";
+        try {
+            apiProjectPath = createProject(openApiFile, apiProjectName, backendCert);
+        } catch (MicroGWTestException e) {
+            if (e.getMessage().equals("Project already exists")) {
+                return Utils.getTargetDirPath() + TestConstant.API_PROJECTS_PATH
+                        + File.separator + apiProjectName + ".zip";
```
Comment (SuKSW) [inline test/test-integration/src/test/java/org/wso2am/micro/gw/tests/util/ApictlUtils.java:31]: Check if zip

File: test/test-integration/src/test/java/org/wso2am/micro/gw/tests/util/ApictlUtils.java [hunk 12]
```diff
@@ -0,0 +1,190 @@
...
+            }
+        } catch (IOException e) {
+            throw new MicroGWTestException("Unable to remove microgateway adapter env from apictl", e);
+        }
+        log.info("Removed apictl microgateway environment: " + mgwEnv);
+    }
+
+    public static void login(String mgwEnv) throws MicroGWTestException {
+        String[] cmdArray = { "mg", "login" };
+        String[] argsArray = { mgwEnv, "-u", "admin", "-p", "admin", "-k" };
```
Comment (SuKSW) [inline test/test-integration/src/test/java/org/wso2am/micro/gw/tests/util/ApictlUtils.java:101]: use cert and remove k

File: test/test-integration/src/test/java/org/wso2am/micro/gw/tests/util/ApictlUtils.java [hunk 13]
```diff
@@ -0,0 +1,190 @@
...
+            if (!responseLines[0].startsWith("Successfully removed environment")) {
+                throw new MicroGWTestException("Unable to remove microgateway adapter env from apictl");
+            }
+        } catch (IOException e) {
+            throw new MicroGWTestException("Unable to remove microgateway adapter env from apictl", e);
+        }
+        log.info("Removed apictl microgateway environment: " + mgwEnv);
+    }
+
+    public static void login(String mgwEnv) throws MicroGWTestException {
```
Comment (SuKSW) [inline test/test-integration/src/test/java/org/wso2am/micro/gw/tests/util/ApictlUtils.java:99]: Add java docs

File: test/test-integration/src/test/resources/openAPIs/deploy_openAPI.yaml [hunk 14]
```diff
@@ -0,0 +1,748 @@
+swagger: '2.0'
```
Comment (SuKSW) [inline test/test-integration/src/test/resources/openAPIs/deploy_openAPI.yaml:1]: Add a swagger 3 file as well

File: test/test-integration/src/test/resources/openAPIs/deploy_openAPI.yaml [hunk 15]
```diff
@@ -0,0 +1,748 @@
...
+        type: array
+        xml:
+          wrapped: true
+        items:
+          type: string
+          xml:
+            name: photoUrl
+      tags:
+        type: array
+        xml:
```
Comment (SuKSW) [inline test/test-integration/src/test/resources/openAPIs/deploy_openAPI.yaml:671]: Add a smaller fike

**Review by Rajith90 — APPROVED** [review]

**Pull request merged**